docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,879 @@
+However, this must anyway be part of their normal reconciliation logic and is a general best practice for Kubernetes controllers.
+
+#### Shoot health checks/conditions
+
+Some of the existing conditions already contain specific code which shall be simplified as well.
+All of the CRDs described above have a `.status.conditions` field to which the controllers may write relevant health information of their function area.
+Gardener will pick them up and copy them over to the Shoots `.status.conditions`.
+
+#### Reconciliation flow
+
+We are now examining the current Shoot creation/reconciliation flow and describe how it would need to be changed when applying this proposal:
+
+| Operation | Description |
+|-----------|-------------|
+| botanist.DeployNamespace | Gardener creates the namespace for the Shoot in the Seed cluster. |
+| botanist.DeployKubeAPIServerService | Gardener creates a Service of type `LoadBalancer` in the Seed.<br>AWS Botanist registers a Mutating Webhook and adds its AWS-specific annotation. |
+| botanist.WaitUntilKubeAPIServerServiceIsReady | Gardener checks the `.status` object of the just created `Service` in the Seed. The contract is that also clouds not supporting load balancers must react on the `Service` object and modify the `.status` to correctly reflect the kube-apiserver's ingress IP. |
+| botanist.DeploySecrets | Gardener creates the secrets/certificates it needs like it does today, but it provides utility functions that can be adopted by Botanists/other controllers if they need additional certificates/secrets created on their own. (We should also add labels to all secrets) |
+| botanist.DeployInternalDomainDNSRecord | Gardener creates a DNS-specific CRD in the Seed, and the responsible DNS-controller picks it up and creates a corresponding DNS record (see CRD specification above). |
+| botanist.DeployExternalDomainDNSRecord | Gardener creates a DNS-specific CRD in the Seed, and the responsible DNS-controller picks it up and creates a corresponding DNS record: (see CRD specification above). |
+| shootCloudBotanist.DeployInfrastructure | Gardener creates a Infrastructure-specific CRD in the Seed, and the responsible Botanist picks it up and does its job: (see CRD above). |
+| botanist.DeployBackupInfrastructure | Gardener creates a `BackupInfrastructure` resource in the Garden cluster.<br>(The BackupInfrastructure controller creates a BackupInfrastructure-specific CRD in the Seed, and the responsible Botanist picks it up and does its job: (see CRD above).) |
+| botanist.WaitUntilBackupInfrastructureReconciled | Gardener checks the `.status` object of the just created `BackupInfrastructure` resource. |
+| hybridBotanist.DeployETCD | Gardener does only deploy the etcd `StatefulSet` without backup-restore sidecar at all.<br>The cloud-specific Botanist registers a Mutating Webhook and adds the backup-restore sidecar, and it also creates the `Secret` needed by the backup-restore sidecar. |
```

Comment:
For the same reasons for which we don't want to introduce the `ControlPlane` CRD (at least not a one having `spec`): Agreed that this may make sense, but let's start easy and let Gardener deploy etcd and the control plane.

Reply:
> We might as well use the opportunity to introduce it here

I'd disagree, the whole implementation will already introduce many changes as it is right now. Introducing an etcd operator at the same time is a high risk and we should first get the other things done without changing too much at the same time.

Reply:
I am not saying we should bring in all the changes in the `etcd` operator right away. It might simply provision the `StatefulSet` as a sidecar just like what's being done right now.

We have to write the `MutatingWebhook` anyway. What's the difference in impact on stability?

Reply:
Every new component is a risk and effort to write. Let's keep the number of the to-be-built controllers small in the first place and start with the most important ones, e.g. those really having logic of one of the two problem dimensions (cloud/operating system).

Reply:
The webhook needs to be written anyway. What is the difference in risk? ;-)

BTW I am not insisting. Just I do not understand.